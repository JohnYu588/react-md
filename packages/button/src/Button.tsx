import {
  RippleContainer,
  useElementInteraction,
  useHigherContrastChildren,
} from "@react-md/core";
import type { ButtonHTMLAttributes } from "react";
import { forwardRef } from "react";
import type { ButtonClassNameThemeOptions } from "./styles";
import { getButtonClassName } from "./styles";

export interface ButtonProps
  extends ButtonHTMLAttributes<HTMLButtonElement>,
    ButtonClassNameThemeOptions {
  /** @defaultValue `"button"` */
  type?: "button" | "reset" | "submit";
}

/**
 * @example
 * Simple Example
 * ```tsx
 * import { Button } from "@react-md/button";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <Button
 *       onClick={(event) => {
 *         // do something
 *       }}
 *     >
 *       Content
 *     </Button>
 *   );
 * }
 * ```
 *
 * @example
 * Theme Example
 * ```tsx
 * import { Button } from "@react-md/button";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <Button
 *       theme="primary"
 *       themeType="contained"
 *       onClick={(event) => {
 *         // do something
 *       }}
 *     >
 *       Content
 *     </Button>
 *   );
 * }
 * ```
 *
 * @example
 * Icon Button Example
 * ```tsx
 * import { Button } from "@react-md/button";
 * import { FavoriteSVGIcon } from "@react-md/material-icons";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <Button
 *       theme="secondary"
 *       themeType="outline"
 *       buttonType="icon"
 *       onClick={(event) => {
 *         // do something
 *       }}
 *     >
 *       <FavoriteSVGIcon />
 *     </Button>
 *   );
 * }
 * ```
 *
 * @example
 * Text Button with icons
 * ```tsx
 * import { Button } from "@react-md/button";
 * import { FavoriteSVGIcon } from "@react-md/material-icons";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <Button
 *       onClick={(event) => {
 *         // do something
 *       }}
 *     >
 *       <FavoriteSVGIcon />
 *       Content
 *     </Button>
 *   );
 * }
 * ```
 */
export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  function Button(props, ref) {
    const {
      type = "button",
      disabled = false,
      theme = "clear",
      themeType = "flat",
      buttonType = "text",
      className,
      children: propChildren,
      onClick,
      onKeyDown,
      onKeyUp,
      onMouseDown,
      onMouseUp,
      onMouseLeave,
      onTouchStart,
      onTouchEnd,
      onTouchMove,
      ...remaining
    } = props;
    const isThemeDisabled = theme === "disabled";
    const { pressed, pressedClassName, rippleContainerProps, handlers } =
      useElementInteraction({
        onClick,
        onKeyDown,
        onKeyUp,
        onMouseDown,
        onMouseUp,
        onMouseLeave,
        onTouchStart,
        onTouchEnd,
        onTouchMove,
        disabled: disabled || isThemeDisabled,
      });

    const children = useHigherContrastChildren(propChildren);

    return (
      <button
        {...remaining}
        // when the theme is set to `"disabled"`, the event handlers should be
        // removed so that it behaves like a disabled button. you do not want to
        // actually set the `disabled` attribute since it will lose keyboard
        // focus. this is mostly for supporting circular progress bars within
        // buttons
        {...(isThemeDisabled ? {} : handlers)}
        aria-disabled={isThemeDisabled || undefined}
        ref={ref}
        type={type}
        className={getButtonClassName({
          theme,
          themeType,
          buttonType,
          disabled,
          pressed,
          pressedClassName,
          className,
        })}
      >
        {children}
        {rippleContainerProps && <RippleContainer {...rippleContainerProps} />}
      </button>
    );
  }
);
