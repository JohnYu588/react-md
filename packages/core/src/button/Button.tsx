"use client";
import { cnb } from "cnbuilder";
import type { ButtonHTMLAttributes } from "react";
import { forwardRef } from "react";
import { RippleContainer } from "../interaction/RippleContainer";
import { useElementInteraction } from "../interaction/useElementInteraction";
import { useHigherContrastChildren } from "../interaction/useHigherContrastChildren";
import type { PropsWithRef } from "../types";
import { bem } from "../utils";
import type {
  FloatingActionButtonPosition,
  FloatingActionButtonProps,
} from "./FloatingActionButton";
import { FloatingActionButton } from "./FloatingActionButton";

declare module "react" {
  interface CSSProperties {
    "--rmd-button-background-color"?: string;
    "--rmd-button-color"?: string;
    "--rmd-button-border-radius"?: string | number;
    "--rmd-button-contained-background-color"?: string;
    "--rmd-button-contained-color"?: string;
    "--rmd-button-text-horizontal-padding"?: string | number;
    "--rmd-button-text-vertical-padding"?: string | number;
    "--rmd-button-text-min-height"?: string | number;
    "--rmd-button-text-min-width"?: string | number;
    "--rmd-button-icon-size"?: string | number;
    "--rmd-button-icon-radius"?: string | number;
  }
}

const styles = bem("rmd-button");

/**
 * When this is set to `"text"`, the size of the button will be determined by
 * the content and will be more block-like. Icons can still be rendered in
 * `"text"` buttons and will have spacing automatically applied between other
 * content in the button.
 *
 * When this is set to `"icon"`, the button will be equal height/width and
 * circular.
 */
export type ButtonType = "text" | "icon" | "icon-square";

/**
 * One of the valid material design default button themes that can be used. This
 * will update the general look and feel by updating the colors within the
 * button while the `ButtonThemeType` will update the borders or box shadow.
 */
export type ButtonTheme =
  | "clear"
  | "primary"
  | "secondary"
  | "warning"
  | "error"
  | "disabled";

/**
 * When this is set to `"flat"`, the button will have no `background-color`,
 * `border`, and `box-shadow`. It will only set the `color` to the
 * {@link ButtonTheme}.
 *
 * When this is set to `"outline"`, the button will have no `background-color`,
 * but gain a `border` and `color` set to the {@link ButtonTheme}.
 *
 * When this is set to `"contained"`, the button will set the `background-color`
 * to the {@link ButtonTheme}, add some `box-shadow`, and set the `color` to
 * either `#000` or `#fff`. (The `color` defaults to whichever value has the
 * highest contrast ratio with the `background-color`)
 */
export type ButtonThemeType = "flat" | "outline" | "contained";

/** @remarks \@since 6.0.0 */
export interface ButtonClassNameThemeOptions {
  className?: string;

  /** @defaultValue `false` */
  disabled?: boolean;

  /**
   * @see {@link ButtonTheme} for information about the different types.
   * @defaultValue `"text"`
   */
  buttonType?: ButtonType;

  /** @defaultValue `"clear"` */
  theme?: ButtonTheme;

  /**
   * @see {@link ButtonThemeType} for information about the theming behavior.
   * @defaultValue `"flat"`
   */
  themeType?: ButtonThemeType;
}

/**
 * @remarks \@since 6.0.0
 * @internal
 */
export interface ButtonClassNameOptions extends ButtonClassNameThemeOptions {
  /** @defaultValue `false` */
  pressed?: boolean;
  pressedClassName?: string;
}

/**
 * Creates a button theme based on the button theming props. This is really just
 * used so that other elements like clickable `<div>`s or `<input type="file">`
 * can look like buttons.
 *
 * @param options - An object containing the themeable button props to generate a
 * button theme className.
 * @returns a string of class names to create an element with a button theme.
 * @remarks \@since 6.0.0 This used to be called `buttonThemeClassNames`.
 */
export function button(options: ButtonClassNameOptions = {}): string {
  const {
    theme: propTheme = "clear",
    themeType = "flat",
    buttonType = "text",
    disabled: propDisabled = false,
    pressed = false,
    pressedClassName,
    className,
  } = options;

  const theme = propTheme === "disabled" ? "clear" : propTheme;
  const disabled = propDisabled || propTheme === "disabled";
  const text = buttonType === "text";
  const icon = !text;
  const outline = themeType === "outline";
  const contained = themeType === "contained";
  const clear = theme === "clear";

  return cnb(
    styles({
      text,
      icon,
      "icon-square": buttonType === "icon-square",
      disabled,
      contained: !disabled && contained,
      outline,
      pressed: contained && pressed,
      [theme]: !disabled && !clear && contained,
      [`text-${theme}`]: !disabled && !clear && !contained,
      [`outline-${theme}`]: !disabled && !clear && outline,
    }),
    pressedClassName,
    className
  );
}

export interface ButtonProps
  extends ButtonHTMLAttributes<HTMLButtonElement>,
    ButtonClassNameThemeOptions {
  /** @defaultValue `"button"` */
  type?: "button" | "reset" | "submit";

  /**
   * The position within the viewport to display the button as a floating action
   * button.
   */
  floating?: FloatingActionButtonPosition;

  /**
   * Any additional props to provide the to `FAB` container element when the
   * `floating` prop is provided
   */
  floatingProps?: PropsWithRef<FloatingActionButtonProps, HTMLSpanElement>;
}

/**
 * **Client Component**
 *
 * @example
 * Simple Example
 * ```tsx
 * import { Button } from "@react-md/core";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <Button
 *       onClick={(event) => {
 *         // do something
 *       }}
 *     >
 *       Content
 *     </Button>
 *   );
 * }
 * ```
 *
 * @example
 * Theme Example
 * ```tsx
 * import { Button } from "@react-md/core";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <Button
 *       theme="primary"
 *       themeType="contained"
 *       onClick={(event) => {
 *         // do something
 *       }}
 *     >
 *       Content
 *     </Button>
 *   );
 * }
 * ```
 *
 * @example
 * Icon Button Example
 * ```tsx
 * import { Button } from "@react-md/core";
 * import { FavoriteSVGIcon } from "@react-md/material-icons";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <Button
 *       theme="secondary"
 *       themeType="outline"
 *       buttonType="icon"
 *       onClick={(event) => {
 *         // do something
 *       }}
 *     >
 *       <FavoriteSVGIcon />
 *     </Button>
 *   );
 * }
 * ```
 *
 * @example
 * Text Button with icons
 * ```tsx
 * import { Button } from "@react-md/core";
 * import { FavoriteSVGIcon } from "@react-md/material-icons";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <Button
 *       onClick={(event) => {
 *         // do something
 *       }}
 *     >
 *       <FavoriteSVGIcon />
 *       Content
 *     </Button>
 *   );
 * }
 * ```
 */
export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  function Button(props, ref) {
    const {
      type = "button",
      disabled = false,
      floating = null,
      floatingProps,
      theme = floating ? "secondary" : "clear",
      themeType = floating ? "contained" : "flat",
      buttonType = floating ? "icon" : "text",
      className,
      children: propChildren,
      onBlur,
      onClick,
      onKeyDown,
      onKeyUp,
      onMouseDown,
      onMouseUp,
      onMouseLeave,
      onDragStart,
      onTouchStart,
      onTouchEnd,
      onTouchMove,
      ...remaining
    } = props;
    const isThemeDisabled = theme === "disabled";
    const { pressed, pressedClassName, rippleContainerProps, handlers } =
      useElementInteraction({
        onBlur,
        onClick,
        onKeyDown,
        onKeyUp,
        onMouseDown,
        onMouseUp,
        onMouseLeave,
        onDragStart,
        onTouchStart,
        onTouchEnd,
        onTouchMove,
        disabled: disabled || isThemeDisabled,
      });

    const children = useHigherContrastChildren(propChildren);

    return (
      <FloatingActionButton position={floating} {...floatingProps}>
        <button
          {...remaining}
          // when the theme is set to `"disabled"`, the event handlers should be
          // removed so that it behaves like a disabled button. you do not want to
          // actually set the `disabled` attribute since it will lose keyboard
          // focus. this is mostly for supporting circular progress bars within
          // buttons
          {...(isThemeDisabled ? {} : handlers)}
          aria-disabled={isThemeDisabled || remaining["aria-disabled"]}
          disabled={disabled}
          ref={ref}
          type={type}
          className={button({
            theme,
            themeType,
            buttonType,
            disabled,
            pressed,
            pressedClassName,
            className,
          })}
        >
          {children}
          {rippleContainerProps && (
            <RippleContainer {...rippleContainerProps} />
          )}
        </button>
      </FloatingActionButton>
    );
  }
);
