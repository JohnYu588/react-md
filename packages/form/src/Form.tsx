import type { FormHTMLAttributes } from "react";
import { forwardRef } from "react";

const noop = (): void => {
  // do nothing
};

export interface FormProps extends FormHTMLAttributes<HTMLFormElement> {
  /**
   * Boolean if the form should no longer prevent default submit behavior. If
   * you enable this prop you should honestly just use a `<form>` element
   * instead
   *
   * @defaultValue `false`
   */
  disablePreventDefault?: boolean;
}

/**
 * This is probably one of the least useful components available as it doesn't
 * do much styling or logic. All this form component will do is add basic flex
 * behavior and prevent the default form submit behavior.
 */
export const Form = forwardRef<HTMLFormElement, FormProps>(function Form(
  props,
  ref
) {
  const {
    children,
    onSubmit = noop,
    disablePreventDefault = false,
    ...remaining
  } = props;

  return (
    <form
      {...remaining}
      onSubmit={(event) => {
        if (!disablePreventDefault) {
          event.preventDefault();
        }

        onSubmit(event);
      }}
      ref={ref}
    >
      {children}
    </form>
  );
});
