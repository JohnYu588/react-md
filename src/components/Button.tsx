import { useElementInteraction } from "@react-md/core";
import { cnb } from "cnbuilder";
import type { ButtonHTMLAttributes, ReactElement, ReactNode } from "react";

import styles from "./Button.module.scss";

export interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  buttonType?: "icon" | "text";
  children: ReactNode;
}

export function Button(props: ButtonProps): ReactElement {
  const {
    type = "button",
    disabled = false,
    className,
    buttonType = "text",
    children,
    ...remaining
  } = props;
  const { pressed, ...handlers } = useElementInteraction({
    disabled,
    ...remaining,
  });

  return (
    <button
      {...remaining}
      {...handlers}
      type={type}
      className={cnb(
        styles.button,
        buttonType === "text" && styles.text,
        buttonType === "icon" && styles.icon,
        pressed && styles.pressed,
        className
      )}
      disabled={disabled}
    >
      {children}
    </button>
  );
}
