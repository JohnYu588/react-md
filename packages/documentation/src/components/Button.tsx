import type { ReactElement, ReactNode } from "react";
import type { ButtonProps } from "@react-md/button";
import { Button as RMDButton } from "@react-md/button";
import { Tooltip, useTooltip } from "@react-md/tooltip";

interface Props extends ButtonProps {
  id: string;
  tooltip?: ReactNode;
}

export default function Button({
  id,
  tooltip,
  children,
  onClick,
  onBlur,
  onFocus,
  onKeyDown,
  onMouseEnter,
  onMouseLeave,
  onTouchStart,
  onContextMenu,
  ...props
}: Props): ReactElement {
  const handlers = {
    onClick,
    onBlur,
    onFocus,
    onKeyDown,
    onMouseEnter,
    onMouseLeave,
    onTouchStart,
    onContextMenu,
  };
  const { elementProps, tooltipProps } = useTooltip({
    baseId: id,
    ...handlers,
  });

  return (
    <>
      <RMDButton {...props} {...(tooltip ? elementProps : { id, ...handlers })}>
        {children}
      </RMDButton>
      <Tooltip {...tooltipProps}>{tooltip}</Tooltip>
    </>
  );
}
