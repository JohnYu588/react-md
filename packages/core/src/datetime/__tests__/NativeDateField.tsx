import { describe, expect, it } from "@jest/globals";
import { type ReactElement, type Ref, createRef } from "react";

import { Button } from "../../button/Button.js";
import { Form } from "../../form/Form.js";
import {
  fireEvent,
  render,
  screen,
  userEvent,
} from "../../test-utils/index.js";
import {
  NativeDateField,
  type NativeDateFieldProps,
} from "../NativeDateField.js";

type TestProps = Partial<NativeDateFieldProps> & {
  inputRef?: Ref<HTMLInputElement>;
};

function Test({
  name = "field",
  inputRef,
  ...options
}: TestProps): ReactElement {
  const { min, max, step, required, disableMessage } = options;
  let messageProps: TestProps["messageProps"];
  if (!disableMessage && (min || max || step || required)) {
    messageProps = {
      "data-testid": "message",
      messageProps: {
        "data-testid": "p",
      },
    };
  }
  return (
    <Form name="form">
      <NativeDateField
        label="Date"
        {...options}
        ref={inputRef}
        name={name}
        containerProps={{
          "data-testid": "container",
        }}
        messageContainerProps={{
          "data-testid": "message-container",
        }}
        messageProps={messageProps}
      />
      <Button type="reset">Reset</Button>
      <Button type="submit">Submit</Button>
    </Form>
  );
}

const setup = (props?: Partial<NativeDateFieldProps>) => {
  const user = userEvent.setup();
  const { rerender } = render(<Test {...props} />);
  const field = screen.getByLabelText("Date");
  const container = screen.getByTestId("container");
  const reset = screen.getByRole("button", { name: "Reset" });
  const submit = screen.getByRole("button", { name: "Submit" });

  return {
    user,
    field,
    container,
    reset,
    submit,
    rerender: (props?: TestProps) => rerender(<Test {...props} />),
  };
};

describe("NativeDateField", () => {
  it("should apply the correct styling, HTML attributes, and allow a ref", () => {
    const ref = createRef<HTMLInputElement>();
    const props = {
      label: "Field",
      name: "field",
      ref,
    } as const;
    const { rerender } = render(<NativeDateField {...props} />);

    const field = screen.getByLabelText("Field");
    expect(ref.current).toBeInstanceOf(HTMLInputElement);
    expect(ref.current).toBe(field);
    expect(field).toMatchSnapshot();

    rerender(
      <NativeDateField
        {...props}
        style={{ color: "white" }}
        className="custom-class-name"
      />
    );
    expect(field).toMatchSnapshot();
  });

  it("should support the readOnly state", () => {
    const { container, field } = setup({ readOnly: true });

    expect(field).toHaveAttribute("readOnly");
    expect(container).toMatchSnapshot();
  });

  it("should support the disabled state", () => {
    const { container, field } = setup({ disabled: true });

    expect(field).toBeDisabled();
    expect(container).toMatchSnapshot();
  });

  it("should support an error state", () => {
    const { container } = setup({ error: true });
    expect(container).toMatchSnapshot();
  });

  it("should not render the FormMessageContainer if the min, max, step, and required props are not provided", () => {
    const { rerender } = setup();

    expect(() => screen.getByTestId("message-container")).toThrow();

    rerender({ min: "2025-01-01", max: "2026-01-01" });
    expect(() => screen.getByTestId("message-container")).not.toThrow();
    expect(() => screen.getByTestId("message")).not.toThrow();

    rerender({ min: "2025-01-01", max: "2026-01-01", disableMessage: true });
    expect(() => screen.getByTestId("message-container")).toThrow();
    expect(() => screen.getByTestId("message")).toThrow();
  });

  it("should handle simple validation inherited from the useTextField hook", async () => {
    const { user, field, container, reset } = setup({
      min: "2025-01-01",
      max: "2026-01-01",
    });
    expect(field).toHaveValue("");
    const message = screen.getByTestId("message");

    // NOTE: Can't actually type like a user for input type="date" at this time.
    await user.type(field, "2025-01-05");
    expect(field).toHaveValue("2025-01-05");
    fireEvent.blur(field);

    expect(message).toBeEmptyDOMElement();
    expect(container).not.toHaveClass("rmd-text-field-container--error");
    expect(message).not.toHaveClass("rmd-error-color");

    await user.clear(field);
    expect(field).toHaveValue("");
    expect(message).toBeEmptyDOMElement();
    expect(container).not.toHaveClass("rmd-text-field-container--error");
    expect(message).not.toHaveClass("rmd-error-color");

    await user.type(field, "2024-04-08{Tab}");
    expect(field).toHaveValue("2024-04-08");
    expect(message).not.toBeEmptyDOMElement();
    expect(container).toHaveClass("rmd-text-field-container--error");
    expect(message).toHaveClass("rmd-error-color");

    await user.click(reset);
    expect(field).toHaveValue("");
    expect(message).toBeEmptyDOMElement();
    expect(container).not.toHaveClass("rmd-text-field-container--error");
    expect(message).not.toHaveClass("rmd-error-color");
  });
});
