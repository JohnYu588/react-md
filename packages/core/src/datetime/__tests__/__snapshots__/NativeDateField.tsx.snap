// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NativeDateField should apply the correct styling, HTML attributes, and allow a ref 1`] = `
<input
  class="rmd-text-field rmd-text-field--placeholder-hidden"
  id="text-field-:r0:"
  name="field"
  placeholder=" "
  type="date"
  value=""
/>
`;

exports[`NativeDateField should apply the correct styling, HTML attributes, and allow a ref 2`] = `
<input
  class="rmd-text-field rmd-text-field--placeholder-hidden"
  id="text-field-:r0:"
  name="field"
  placeholder=" "
  type="date"
  value=""
/>
`;

exports[`NativeDateField should support an error state 1`] = `
<div
  class="rmd-text-field-container rmd-text-field-container--outline rmd-text-field-container--hoverable rmd-text-field-container--label"
  data-testid="container"
>
  <input
    class="rmd-text-field rmd-text-field--placeholder-hidden"
    id="text-field-:r6:"
    name="field"
    placeholder=" "
    type="date"
    value=""
  />
  <label
    class="rmd-label rmd-label--floating"
    for="text-field-:r6:"
  >
    Date
  </label>
</div>
`;

exports[`NativeDateField should support the disabled state 1`] = `
<div
  class="rmd-text-field-container rmd-text-field-container--outline rmd-text-field-container--disabled rmd-text-field-container--label rmd-text-disabled-color"
  data-testid="container"
>
  <input
    class="rmd-text-field rmd-text-field--placeholder-hidden"
    disabled=""
    id="text-field-:r4:"
    name="field"
    placeholder=" "
    type="date"
    value=""
  />
  <label
    class="rmd-label rmd-label--disabled rmd-label--floating rmd-text-disabled-color"
    for="text-field-:r4:"
  >
    Date
  </label>
</div>
`;

exports[`NativeDateField should support the readOnly state 1`] = `
<div
  class="rmd-text-field-container rmd-text-field-container--outline rmd-text-field-container--disabled rmd-text-field-container--hoverable rmd-text-field-container--label"
  data-testid="container"
>
  <input
    class="rmd-text-field rmd-text-field--placeholder-hidden"
    id="text-field-:r2:"
    name="field"
    placeholder=" "
    readonly=""
    type="date"
    value=""
  />
  <label
    class="rmd-label rmd-label--floating"
    for="text-field-:r2:"
  >
    Date
  </label>
</div>
`;
